cluster mempool: add TxGraph reorg functionality

sipa opened the pull request:
Part of cluster mempool (#30289).

During reorganisations, it is possible that dependencies get added which would result in clusters that violate policy limits (cluster count, cluster weight), when linking the new from-block transactions to the old from-mempool transactions. Unlike RBF scenarios, we cannot simply reject the changes when they are due to received blocks. To accommodate this, add a `TxGraph::Trim()`, which removes some subset of transactions (including descendants) in order to make all resulting clusters satisfy the limits.

Conceptually, the way this is done is by defining a rudimentary linearization for the entire would-be too-large cluster, iterating it from beginning to end, and reasoning about the counts and weights of the clusters that would be reached using transactions up to that point. If a transaction is encountered whose addition would violate the limit, it is removed, together with all its descendants.

This rudimentary linearization is like a merge sort of the chunks of the clusters being combined, but respecting topology. More specifically, it is continuously picking the highest-chunk-feerate remaining transaction among those which have no unmet dependencies left. For efficiency, this rudimentary linearization is computed lazily, by putting all viable transactions in a heap, sorted by chunk feerate, and adding new transactions to it as they become viable.

The `Trim()` function is rather unusual compared to the `TxGraph` functionality added in previous PRs, in that `Trim()` makes it own decisions about what the resulting graph contents will be, without good specification of how it makes that decision - it is just a best-effort attempt (which is improved in the last commit). All other `TxGraph` mutators are simply to inform the graph about changes the calling mempool code decided on; this one lets the decision be made by txgraph.

As part of this, the "oversized" property is expanded to also encompass a configurable cluster weight limit (in addition to cluster count limit).

sipa committed:
txgraph: Add ability to trim oversized clusters (feature)

During reorganisations, it is possible that dependencies get add which
result in clusters that violate limits (count, size), when linking the
new from-block transactions to the old from-mempool transactions.

Unlike RBF scenarios, we cannot simply reject these policy violations
when they are due to received blocks. To accomodate this, add a Trim()
function to TxGraph, which removes transactions (including descendants)
in order to make all resulting clusters satisfy the limits.

In the initial version of the function added here, the following approach
is used:
- Lazily compute a naive linearization for the to-be-merged cluster (using
  an O(n log n) algorithm, optimized for far larger groups of transactions
  than the normal linearization code).
- Initialize a set of accepted transactions to {}
- Iterate over the transactions in this cluster one by one:
  - If adding the transaction to the set makes it exceed the max cluster size
    or count limit, stop.
  - Add the transaction to the set.
- Remove all transactions from the cluster that were not included in the set
  (note that this necessarily includes all descendants too, because they
  appear later in the naive linearization).

Co-authored-by: Greg Sanders <gsanders87@gmail.com>

sipa committed:
txgraph: Add ability to configure maximum cluster size/weight (feature)

This is integrated with the oversized property: the graph is oversized when
any connected component within it contains more than the cluster count limit
many transactions, or when their combined size/weight exceeds the cluster size
limit.

It becomes disallowed to call AddTransaction with a size larger than this limit,
though this limit will be lifted in the next commit.

In addition, SetTransactionFeeRate becomes SetTransactionFee, so that we do not
need to deal with the case that a call to this function might affect the
oversizedness.

sipa committed:
txgraph: Track multiple potential would-be clusters in Trim (improvement)

In the existing Trim function, as soon as the set of accepted transactions
would exceed the max cluster size or count limit, the acceptance loop is
stopped, removing all later transactions. However, it is possible that by
excluding some of those transactions the would-be cluster splits apart into
multiple would-clusters. And those clusters may well permit far more
transactions before their limits are reached.

Take this into account by using a union-find structure inside TrimTxData to
keep track of the count/size of all would-be clusters that would be formed
at any point, and only reject transactions which would cause these resulting
partitions to exceed their limits.

This is not an optimization in terms of CPU usage or memory; it just
improves the quality of the transactions removed by Trim().

sipa committed:
txgraph: add Trim benchmark (benchmark)

theuni reviewed (commented):
Very quick and shallow pass through the initial impl commit. This PR is a lot to get through :)

theuni commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,1193 @@
+
+TxGraph::Ref& TxGraph::Ref::operator=(Ref&& other) noexcept
+{
+    // Inform both TxGraphs about the Refs being swapped.
```
> > Why is this doing an effective swap?
> 
> I think this is quite common, that move-construction is effectively performing a swap.
> 
> > I would expect this to call `UnlinkRef` on the moved-from value and reset its `m_graph` and `m_index`
> 
> That's possible too, and slightly more efficient I guess.
> 
> > Otherwise it wouldn't be unlinked until the moved-from variable goes out of scope, no?
> 
> Indeed. I don't think that's a problem.

Afaik the move/swap idiom is only safe if the swapped-to value's dtor doesn't have any interesting ordering requirements or side-effects.

As a contrived example, a user may do something like:
```c++
std::vector<TxGraph::Ref> vec;

vec.push_back(txgraph->AddTransaction(fee));
auto ref = txgraph->AddTransaction(fee2);
...
ref = std::move(vec.back());
```
The vector now holds the old ref and `UnlinkRef` will not be called until that element is removed. I realize it's allowed to be a "valid but unspecified state", but I wouldn't expect a ref to be hanging around.

theuni commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,1193 @@
+    auto idx = m_entries.size();
+    m_entries.emplace_back();
+    auto& entry = m_entries.back();
+    entry.m_ref = &ret;
```
Ah, right, I missed that the move ctor would handle the update. Thanks for explaining.

sdaftuar commented on src/txgraph.cpp:
```diff
@@ -1449,6 +1465,7 @@ Cluster::Cluster(TxGraphImpl& graph, const FeeFrac& feerate, GraphIndex graph_in
 TxGraph::Ref TxGraphImpl::AddTransaction(const FeeFrac& feerate) noexcept
 {
     Assume(m_chunkindex_observers == 0 || m_clustersets.size() > 1);
+    Assume(feerate.size > 0 && uint64_t(feerate.size) <= m_max_cluster_size);
```
FYI -- in my rebase of #28676, I'm seeing `tx_pool` fuzz test failures due to this line. Not clear to me whether we should require the caller to enforce the policy requirement that a single tx be below the cluster size limit, or just let the caller discover a changeset is oversized and then reject?

sipa committed:
txgraph: Permit transactions that exceed cluster size limit (feature)

This removes the restriction added in the previous commit that individual
transactions do not exceed the max cluster size limit.

With this change, the responsibility for enforcing cluster size limits can
be localized purely in TxGraph, without callers (and in particular, tests)
needing to duplicate the enforcement for individual transactions.

yancyribbens commented on src/cluster_linearize.h:
```diff
@@ -1336,6 +1347,38 @@ std::vector<ClusterIndex> MergeLinearizations(const DepGraph<SetType>& depgraph,
+        // Find which position to place elem in (updating j), continuously moving the elements
+        // in between forward.
+        while (place_before.Any()) {
+            // j cannot be 0 here; if it was, then there was necessarily nothing earlier which
```
```suggestion
            // j cannot be 0 or less; if it was, then there was necessarily nothing earlier which
```

yancyribbens replied:
```
if it was, then there was necessarily nothing earlier which elem needs to be place before anymore, and place_before would be empty.
```
This comment seems jumbled and hard to understand.  Is it possible to word this better?

yancyribbens commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,2801 @@
+{
+    auto& entry = m_entries[idx];
+    if (!m_chunkindex_discarded.empty()) {
+        // Reuse an discarded node handle.
```
```suggestion
        // Reuse a discarded node handle.
```

instagibbs commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -788,6 +788,27 @@ FUZZ_TARGET(txgraph)
                     assert(sum == worst_chunk_feerate);
                 }
                 break;
+            } else if ((block_builders.empty() || sims.size() > 1) && command-- == 0) {
```
Here are a couple suggestions to tighten up the checks on what we're trimming. Attempted to state that:
1) non-oversized clusters remain untouched
2) transactions with no ancestors that are undersized should never be trimmed (this seems natural but the interface doesn't guarantee it)

Mostly trying to avoid degenerate behavior like entire mempool being wiped, or any oversized cluster being completely wiped needlessly.
```
diff --git a/src/test/fuzz/txgraph.cpp b/src/test/fuzz/txgraph.cpp
index 47ad5792bf..bdf759f485 100644
--- a/src/test/fuzz/txgraph.cpp
+++ b/src/test/fuzz/txgraph.cpp
@@ -91,4 +91,28 @@ struct SimTxGraph
     }
 
+    /** Returns true if all positions in the given set are in oversized clusters, false otherwise. */
+    bool AllInOversizedClusters(SetType set)
+    {
+        if (set.Any() && !IsOversized()) return false;
+
+        auto todo = graph.Positions();
+        if (!set.IsSubsetOf(todo)) return false;
+
+        // Walk all clusters, and make sure all of `set` doesn't come from non-oversized clusters
+        while (todo.Any()) {
+            auto component = graph.FindConnectedComponent(todo);
+            bool is_oversized{component.Count() > max_cluster_count};
+            uint64_t component_size{0};
+            for (auto i : component) component_size += graph.FeeRate(i).size;
+            is_oversized |= component_size > max_cluster_size;
+            // Some element existed in a non-oversized cluster
+            if (!is_oversized && set.Overlaps(component)) {
+                return false;
+            }
+            todo -= component;
+        }
+        return true;
+    }
+
     void MakeModified(DepGraphIndex index)
     {
@@ -798,8 +822,16 @@ FUZZ_TARGET(txgraph)
                 }
                 auto removed_set = top_sim.MakeSet(removed);
-                // The removed set must contain all its own descendants.
                 for (auto simpos : removed_set) {
+                    // The removed set must contain all its own descendants.
                     assert(top_sim.graph.Descendants(simpos).IsSubsetOf(removed_set));
+
+                    // Nothing removed should have no ancestors and be undersized itself
+                    assert(!(top_sim.GetAncDesc(top_sim.GetRef(simpos), /*desc=*/false).Count() == 0 &&
+                        (uint64_t) top_sim.graph.FeeRate(simpos).size <= top_sim.max_cluster_size));
                 }
+
+                // Nothing from not-oversized-clusters should have been removed
+                assert(top_sim.AllInOversizedClusters(removed_set));
+
                 // Apply all removals to the simulation, and verify the result is no longer
                 // oversized. Don't query the real graph for oversizedness; it is compared
```

instagibbs reviewed (commented):
acb2b6a90a47567773268a024bcf2f1ddfcc60df

focused on motivation/structure/approach and test coverage for first pass. Haven't validated the actual trimming logic yet.

instagibbs committed:
txgraph: remove unnecessary m_group_oversized (simplification)

ismaelsadeeq reviewed (commented):
Code review up to 412a9e0e69ccf461554660b7c622b768cfb8ccb3

Also did some fuzzing and mutate the new code added and tests, craches occur as expected.

> If a transaction is encountered whose addition would violate the limit, it is removed, together with all its descendants.

In real-world scenarios it is unlikely we encounter those transactions that will exceed the cluster count limit, however when it occur I assume that the newly connected block will have similar transaction with the reorged block. I think maintaining an in-memory data structure for these transactions removed due to the limit seperate from mempool will reduce bandwidth from re-questing the tx again and also avoid non-contextual checks?

We can get rid of them after the reorg is complete and they arent included in any block.

sipa reviewed (commented):
> In real-world scenarios it is unlikely we encounter those transactions that will exceed the cluster count limit, however when it occur I assume that the newly connected block will have similar transaction with the reorged block. I think maintaining an in-memory data structure for these transactions removed due to the limit seperate from mempool will reduce bandwidth from re-questing the tx again and also avoid non-contextual checks?

In theory, almost the entire mempool could be trimmed away (in a pathological case where the re-added block transactions link all mempool transaction together), so any approach for doing this would need to be able to handle picking just an acceptable-sized subset. It may make sense to just throw them into the compact block reconstruction extrapool though, up to a certain limit.

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -101,6 +104,9 @@ class Cluster
     }
+    /** Whether this cluster is oversized (just due to the size of its transaction(s), not due to
+     *  dependencies that are yet to be added. */
+    bool IsOversized() const noexcept { return m_quality == QualityLevel::OVERSIZED_SINGLETON; }
```
> To be clear, Cluster::IsOversized() returns whether the Cluster is oversized in general. 

huh, I missed thanks for the clarity.
I think you can add this comment as well  or anything that would make this explicit and clearer will be appreciated.

ismaelsadeeq reviewed (commented):
light code review ACK 538e9ff804f8dfba6f6a808e83572fdeab145ab8

instagibbs reviewed (commented):
Reviewed through 538e9ff804f8dfba6f6a808e83572fdeab145ab8

Last big Q is about the cycles being generated in the implied graph, and what if anything we should do differently about it.

sipa committed:
txgraph: add fuzz test scenario that avoids cycles inside Trim() (tests)

Trim internally builds an approximate dependency graph of the merged cluster,
replacing all existing dependencies within existing clusters with a simple
linear chain of dependencies. This helps keep the complexity of the merging
operation down, but may result in cycles to appear in the general case, even
though in real scenarios (where Trim is called for stitching re-added mempool
transactions after a reorg back to the existing mempool transactions) such
cycles are not possible.

Add a test that specifically targets Trim() but in scenarios where it is
guaranteed not to have any cycles. It is a special case, is much more a
whitebox test than a blackbox test, and relies on randomness rather than
fuzz input. The upside is that somewhat stronger properties can be tested.

Co-authored-by: Greg Sanders <gsanders87@gmail.com>

ismaelsadeeq commented:
reACK e1f501cab224608004e3a1eee7a48eec3cea25fc [deab145ab8..e1f501cab2](https://github.com/bitcoin/bitcoin/compare/538e9ff804f8dfba6f6a808e83572fdeab145ab8..e1f501cab224608004e3a1eee7a48eec3cea25fc)

Following the last light ACK, review comments by @instagibbs were addressed:
- Increased fuzz test coverage
- Clearer and corrected comments
- An Assume statement was added

glozow reviewed (commented):
Small doc things for now, still reviewing

sipa reviewed (commented):
Addressed nit, and improved comments here and there.

Also two bigger changes:
* There was a bug in the second Trim commit (the one that added UF), resulting in it not achieving the desired quality (if anyone has ideas for tests that exercise the improved quality UF brings, I'm all ears).
* I've changed the conversion of unincluded transactions to singleton UF partitions to happen at the time they are *popped* from the heap, rather than when they are added. I think this makes more sense, as it means the heap is purely for unincluded but includable transactions, and the UF structure purely for included or rejected transactions.

---

Effect of the bugfix, in a benchmark that creates a merge-the-whole-mempool scenario (not included in the PR, because it doesn't really fit in our bench framework):

Before fix:
* Benchmarks (in ms):
  * block_build: 21.7986
  * oversize: 7.12257
  * block_linearize: 33.0116
  * mempool_build: 0.0691267
  * trim: 10.4955
  * final_linearize: 1.5371
  * reset: 0.308329
* Transaction counts:
  * before trim: 64011
  * after trim: 38743

After fix:
* Benchmarks (in ms):
  * block_build: 23.3451
  * oversize: 7.70649
  * block_linearize: 34.05
  * mempool_build: 0.0766003
  * trim: 12.6596
  * final_linearize: 0.0175771
  * reset: 0.371984
* Transaction counts:
  * before trim: 64011
  * after trim: 63994

glozow committed:
txgraph: add unit test for TxGraph::Trim (tests)

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

glozow reviewed (commented):
> if anyone has ideas for tests that exercise the improved quality UF brings, I'm all ears

Maybe a starting point, I wrote a unit test to see the difference in trimming with 0b59916271c43cbedd14aafe88e5baf2dec2d97d: https://github.com/glozow/bitcoin/commits/2025-06-31553-test/

One test is a bunch of singleton transactions (e.g. from disconnected blocks) connected by 1 transaction (e.g. in the mempool).
- Initial `Trim` implementation: trims all but the top `max_cluster_count` transactions of the parents
- With union-find tracking multiple: just trims the child, which leaves everything as singleton clusters

The other test is a trellis (e.g. parents all from disconnected blocks, children all already in mempool)
- Initial `Trim` implementation: trims half of the trellis (top and bottom), leaving just the prefix that is within `max_cluster_count` size
- With union-find tracking multiple: just trims the middle transaction, splitting it into 2 trellises

> There was a bug in the second Trim commit

I haven't fully grokked this commit (there's something wrong with my eyesight whenever I look at the code), but before the last push, the trellis test was evicting 1/4 of the transactions: https://github.com/glozow/bitcoin/commits/2025-06-31553-trellis/

glozow commented on src/txgraph.h:
```diff
@@ -240,8 +240,9 @@ class TxGraph
-/** Construct a new TxGraph with the specified limit on transactions within a cluster. That
- *  number cannot exceed MAX_CLUSTER_COUNT_LIMIT. */
-std::unique_ptr<TxGraph> MakeTxGraph(unsigned max_cluster_count) noexcept;
+/** Construct a new TxGraph with the specified limit on the number of transactions within a cluster,
```
Fair enough, I suppose they are both external policy limits!

instagibbs commented:
@sipa to be clear, was the issue 
```
    -+            if (new_count > m_max_cluster_count || new_size > m_max_cluster_size) break;
    ++            if (new_count > m_max_cluster_count || new_size > m_max_cluster_size) continue;
```   

In that the optimization means we should keep on going in case we're going to create/add to undersized clusters, but this `break` broke the optimization. In other words, the base performance was still intact?

instagibbs reviewed (approved):
reACK 643fffafdc70da5fbaf1992c36965595b9e63f84

> and the UF structure purely for included or rejected transactions

i.e. rejected here means "things with no unmet deps yet never included"

Validated in my mental model that bug fix and heap logic changes are correct. Reviewed new unit tests and makes sense that it would have caught the bug. It's a bit further non-black-box, but they're simple.

`git range-diff master e1f501cab224608004e3a1eee7a48eec3cea25fc 643fffafdc70da5fbaf1992c36965595b9e63f84`

instagibbs commented:
reACK 2b2df98747fdb6380588991167ce2e8cb92f3bfb

instagibbs commented:
@sipa linked here, added to follow-up potentially?

instagibbs commented:
@sipa one order of magnitude more for me on my wimpy laptop but well within acceptable bounds I suspect for a reorg. IIUC this is one of the more pessimistic mempool topologies for this since you're going to have to go through the ~entire mempool before finishing

glozow commented on src/txgraph.cpp:
```diff
@@ -2466,6 +2602,255 @@ std::pair<std::vector<TxGraph::Ref*>, FeePerWeight> TxGraphImpl::GetWorstMainChu
+            trim_it->m_parent_count = trim_it->m_deps_left;
+            // If this transaction has no unmet dependencies, and is not oversized, add it to the
+            // heap (just append for now, the heapification happens below).
+            if (trim_it->m_deps_left == 0 && trim_it->m_tx_size <= m_max_cluster_size) {
```
I tested that `Trim()` always removes oversized singletons, even if there are no clusters to be merged, using a simple unit test

```
diff --git a/src/test/txgraph_tests.cpp b/src/test/txgraph_tests.cpp
index 9f9ad04ff37..d90afd8819d 100644
--- a/src/test/txgraph_tests.cpp
+++ b/src/test/txgraph_tests.cpp
@@ -248,4 +248,43 @@ BOOST_AUTO_TEST_CASE(txgraph_trim_huge)
     BOOST_CHECK(graph->GetTransactionCount() >= (NUM_TOP_CHAINS * NUM_TX_PER_TOP_CHAIN * 99) / 100);
 }
 
+BOOST_AUTO_TEST_CASE(txgraph_trim_big_singletons)
+{
+    // Mempool consists of 100 singleton clusters; there are no dependencies. Some are oversized. Trim() should remove all of the oversized ones.
+    static constexpr int MAX_CLUSTER_COUNT = 64;
+    static constexpr int32_t MAX_CLUSTER_SIZE = 100'000;
+    static constexpr int NUM_TOTAL_TX = 100;
+
+    // Create a new graph for the test.
+    auto graph = MakeTxGraph(MAX_CLUSTER_COUNT, MAX_CLUSTER_SIZE);
+
+    // Add all transactions and store their Refs.
+    std::vector<TxGraph::Ref> refs;
+    refs.reserve(NUM_TOTAL_TX);
+
+    // Add all transactions. They are in individual clusters.
+    for (unsigned int i = 0; i < NUM_TOTAL_TX; ++i) {
+        // The 88th transaction is oversized.
+        // Every 20th transaction is oversized.
+        const FeePerWeight feerate{500 + i, (i == 88 || i % 20 == 0) ? MAX_CLUSTER_SIZE + 1 : 100};
+        refs.push_back(graph->AddTransaction(feerate));
+    }
+    graph->SanityCheck();
+
+    // Check that the graph is now oversized. This also forces the graph to
+    // group clusters and compute the oversized status.
+    BOOST_CHECK(graph->IsOversized(/*main_only=*/false));
+
+    // Call Trim() to remove transactions and bring the cluster back within limits.
+    auto removed_refs = graph->Trim();
+    graph->SanityCheck();
+    BOOST_CHECK_EQUAL(graph->GetTransactionCount(), NUM_TOTAL_TX - 6);
+    BOOST_CHECK(!graph->IsOversized(/*main_only=*/false));
+
+    // Check that all the oversized transactions were removed.
+    for (unsigned int i = 0; i < refs.size(); ++i) {
+        BOOST_CHECK_EQUAL(graph->Exists(refs[i]), i != 88 && i % 20 != 0);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/txgraph.cpp b/src/txgraph.cpp
index 1b29a05e534..26eef54d321 100644
--- a/src/txgraph.cpp
+++ b/src/txgraph.cpp
@@ -2745,7 +2745,7 @@ std::vector<TxGraph::Ref*> TxGraphImpl::Trim() noexcept
             trim_it->m_parent_count = trim_it->m_deps_left;
             // If this transaction has no unmet dependencies, and is not oversized, add it to the
             // heap (just append for now, the heapification happens below).
-            if (trim_it->m_deps_left == 0 && trim_it->m_tx_size <= m_max_cluster_size) {
+            if (trim_it->m_deps_left == 0) {
                 trim_heap.push_back(trim_it);
             }
         }
@@ -2813,7 +2813,7 @@ std::vector<TxGraph::Ref*> TxGraphImpl::Trim() noexcept
                 new_size += ptr->m_uf_size;
             }
             // Skip the entry if this would violate any limit.
-            if (new_count > m_max_cluster_count || new_size > m_max_cluster_size) continue;
+            if (new_count > m_max_cluster_count) continue;
 
             // Union the partitions this transaction and all its dependencies are in together.
             auto rep = &entry;
```

glozow reviewed (commented):
ACK 2b2df98747fdb6380588991167ce2e8cb92f3bfb

Reviewed the code and the tests. I ran the bench to see that we spend a lot longer on the last commit since we are able to consider more potential clusters.
Before:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       55,112,375.00 |               18.14 |    0.0% |      0.06 | `TxGraphTrim`
```
After:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       71,068,959.00 |               14.07 |    0.0% |      0.07 | `TxGraphTrim`
```

instagibbs reviewed (approved):
reACK 1632fc104be8f171f59a023800b2f9f20d0a3cff

Includes the benchmark and unit test for trimming singletons, and a text cleanup

`git range-diff master 2b2df98747fdb6380588991167ce2e8cb92f3bfb 1632fc104be8f171f59a023800b2f9f20d0a3cff`

glozow commented:
reACK 1632fc104be via range-diff

glozow reviewed (commented):
reACK 1632fc104be via range-diff

ismaelsadeeq reviewed (approved):
reACK 1632fc104be8f171f59a023800b2f9f20d0a3cff 🛰️
